diff --git a/router/pppd/pppd/lcp.c b/router/pppd/pppd/lcp.c
index d4bb016..a053255 100644
--- pppd/pppd/lcp.c
+++ ppp-2.4.1.pppoe4.orig/pppd/lcp.c
@@ -2110,24 +2110,6 @@ void LcpLinkFailure (f)
 }

 /*
- *  Check if we have been received "echo reply" packet
- */
-
-static void LcpCheckReply(arg)
-       void *arg;
-{
-       //fsm *f1 = &lcp_fsm[ifunit];
-       //junzhao 2004.7.9
-       fsm *f1 = (fsm *)arg;
-
-       if(lcp_echos_pending >= 1)
-               TIMEOUT(LcpEchoTimeout, f1, 1);
-       else
-               TIMEOUT(LcpEchoTimeout, f1, lcp_echo_interval-1);
-}
-
-
-/*
  * Timer expired for the LCP echo requests from this process.
  */

@@ -2135,9 +2117,7 @@ static void
 LcpEchoCheck (f)
     fsm *f;
 {
-    //junzhao 2004.7.9
-    if(check_adsl_status())
-         LcpSendEchoRequest (f);
+    LcpSendEchoRequest (f);
     if (f->state != OPENED)
        return;

@@ -2146,9 +2126,7 @@ LcpEchoCheck (f)
      */
     if (lcp_echo_timer_running)
        warn("assertion lcp_echo_timer_running==0 failed");
-    //TIMEOUT (LcpEchoTimeout, f, lcp_echo_interval);
-    //junzhao 2004.3.18
-    TIMEOUT (LcpCheckReply, f, 1);
+    TIMEOUT (LcpEchoTimeout, f, lcp_echo_interval);
     lcp_echo_timer_running = 1;
 } 
